### PR TITLE
fallback to packaged global db when resolving assets

### LINKED ISSUE #5046: Fallback to shipped DB for constant assets if missing from globalDB
## Problem Definition

We hit a lot of problems with the new approach of storing assets in the DB without having them defined as constants in the codebase anymore. An example can be seen here: https://github.com/rotki/rotki/pull/5045

We need to have a way to allows us to have guaranteed access to a set of assets .... even if the user's global DB is very very old... even if the user went and deleted them himself

## Task

Need to think but I guess the best idea is to always fallback to the shipped DB for a set of assets (let's call them the constant assets).

Other ideas also welcome.

### PR DESCRIPTION
**Pull request by prettyirrelevant**
Closes #5046

### CODE REVIEW

File: rotkehlchen/globaldb/handler.py [hunk 1]
```diff
@@ -627,12 +627,19 @@ def get_asset_data(
...
+            from_builtin_db: bool = False,
     ) -> Optional[List[UnderlyingToken]]:
         """Fetch underlying tokens for a token address if they exist"""
-        cursor.execute(
-            'SELECT B.address, B.token_kind, A.weight FROM underlying_tokens_list AS A JOIN evm_tokens as B WHERE A.identifier=B.identifier AND parent_token_entry=?;',  # noqa: E501
-            (parent_token_identifier,),
-        )
+        if from_builtin_db is True:
+            cursor.execute(
+                'SELECT B.address, B.token_kind, A.weight FROM builtin_db.underlying_tokens_list AS A JOIN builtin_db.evm_tokens as B WHERE A.identifier=B.identifier AND parent_token_entry=?;',  # noqa: E501
```
Comment (nebolax) [inline rotkehlchen/globaldb/handler.py:635]: Maybe instead of separating it into 2 completely different queries, just make one parametrizable?

**Review by nebolax — COMMENTED** [review]
Also I think it would be good to have tests

File: rotkehlchen/globaldb/handler.py [hunk 2]
```diff
@@ -1765,7 +1772,10 @@ def resolve_asset(identifier: str, form_with_incomplete_data: bool) -> AssetWith
 
             asset_data = cursor.fetchone()
             if asset_data is None:
-                raise UnknownAsset(identifier)
+                return GlobalDBHandler()._resolve_asset_from_packaged_db(
```
[the remark this answers is not in the document]
Comment (LefterisJP) [reply rotkehlchen/globaldb/handler.py:1775]: That is absolutely a terrible idea. Read the issue @prettyirrelevant. It should only happen when the asset is in the list of constant assets and an unknown asset or a wrong asset type exception is raised

File: rotkehlchen/globaldb/handler.py [hunk 3]
```diff
@@ -1765,7 +1772,10 @@ def resolve_asset(identifier: str, form_with_incomplete_data: bool) -> AssetWith
...
+                    identifier=identifier,
```
[the remark this answers is not in the document]
Comment (LefterisJP) [reply rotkehlchen/globaldb/handler.py:1776]: It should not be attached an unattached each time. This should only happen in a very very rare case, which is the user deleting an asset that is in packaged DB and is used as a constant asset in the codebase

File: rotkehlchen/globaldb/handler.py [hunk 4]
```diff
@@ -1801,3 +1811,62 @@ def get_asset_type(identifier: str) -> AssetType:
...
+
+        May raise:
+        - UnknownAsset
+        """
+        log.debug(f'Attempt to resolve asset {identifier} using the packaged database')
+        root_dir = Path(__file__).resolve().parent.parent
+        builtin_database = root_dir / 'data' / 'global.db'
+
+        query = """
+        SELECT A.identifier, A.type, B.address, B.decimals, A.name, C.symbol, C.started, null, C.swapped_for, C.coingecko, C.cryptocompare, B.protocol, B.chain, B.token_kind, null, null FROM builtin_db.assets as A JOIN builtin_db.evm_tokens as B
```
Comment (yabirgb) [inline rotkehlchen/globaldb/handler.py:1832]: Don't duplicate code, it should reuse the same logic as it is the same database schema

File: docs/changelog.rst [hunk 5]
```diff
@@ -1,7 +1,7 @@
 =========
 Changelog
 =========
-
+* :feature:`5046` Improve experience for users with old database when resolving assets.
```
Comment (yabirgb) [inline docs/changelog.rst:4]: Add a test for this adding an asset that should be constant and fails to resolve so it queries it from the packaged globaldb. I believe is enough if you delete LUSD for example from the db

File: rotkehlchen/globaldb/handler.py [hunk 6]
```diff
@@ -1801,3 +1811,62 @@ def get_asset_type(identifier: str) -> AssetType:
             raise UnknownAsset(identifier)
```
[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/globaldb/handler.py:1811]: Yes. For the logic to be consistent and the behavior to work as expected it should be also modified

File: rotkehlchen/globaldb/handler.py [hunk 7]
```diff
@@ -627,12 +627,19 @@ def get_asset_data(
...
+        if from_builtin_db is True:
+            cursor.execute(
+                'SELECT B.address, B.token_kind, A.weight FROM builtin_db.underlying_tokens_list AS A JOIN builtin_db.evm_tokens as B WHERE A.identifier=B.identifier AND parent_token_entry=?;',  # noqa: E501
+                (parent_token_identifier,),
+            )
+        else:
+            cursor.execute(
+                'SELECT B.address, B.token_kind, A.weight FROM underlying_tokens_list AS A JOIN evm_tokens as B WHERE A.identifier=B.identifier AND parent_token_entry=?;',  # noqa: E501
+                (parent_token_identifier,),
+            )
```
Comment (yabirgb) [inline rotkehlchen/globaldb/handler.py:642]: I believe you don't need to attach it. Just changing the cursor you use should be enough. The performance of the database can be affected.

edit: I searched and found this comment regarding the topic https://stackoverflow.com/a/50782262

**Review by yabirgb — COMMENTED** [review]
Synced with @prettyirrelevant about the PR. Leaving some more comments

File: rotkehlchen/tests/unit/test_globaldb.py [hunk 8]
```diff
@@ -988,3 +988,15 @@ def test_general_cache(globaldb):
...
+def test_packaged_db_check_for_constant_assets(globaldb):
+    """Check that if UnknownAsset is raised for asset in CONSTANT_ASSETS, check the packaged db"""
+
+    # delete one entry in `CONSTANT_ASSETS`
+    with globaldb.conn.write_ctx() as cursor:
+        cursor.execute('DELETE FROM assets WHERE identifier=?', (A_LUSD.identifier,))
+        assert cursor.rowcount == 1
+
+    # now resolve the asset and see that no error is raised
+    assert A_LUSD.resolve_to_crypto_asset()
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/test_globaldb.py:1002]: Also try an asset that doesn't exists in any of the two globaldbs

File: rotkehlchen/globaldb/handler.py [hunk 9]
```diff
@@ -135,6 +136,19 @@ def __new__(
         GlobalDBHandler.__instance.conn = _initialize_global_db_directory(data_dir, sql_vm_instructions_cb)  # noqa: E501
         return GlobalDBHandler.__instance
 
+    @property
+    def packaged_db_conn(self) -> DBConnection:
```
Comment (yabirgb) [inline rotkehlchen/globaldb/handler.py:140]: We prefer against using properties because they tend to hide code complexity. If you see it might look like a normal attribute but it could have slow logic behind

File: rotkehlchen/globaldb/handler.py [hunk 10]
```diff
@@ -135,6 +136,19 @@ def __new__(
...
+        """Return a `DBConnection` instance for the packaged global db."""
```
Comment (yabirgb) [inline rotkehlchen/globaldb/handler.py:141]:
```suggestion
        """Return a DBConnection instance for the packaged global db."""
```

File: rotkehlchen/assets/resolver.py [hunk 11]
```diff
@@ -105,15 +121,31 @@ def resolve_asset_to_class(
...
+            resolved_asset = GlobalDBHandler().resolve_asset(
                 identifier=identifier,
-                expected_type=expected_type,
-                real_type=type(resolved_asset),
+                form_with_incomplete_data=form_with_incomplete_data,
+                use_packaged_db=True,
             )
-        # resolve_asset returns Asset but we already narrow type with the if check above
-        return resolved_asset  # type: ignore
+            AssetResolver().assets_cache.set(identifier, resolved_asset)
```
Comment (yabirgb) [inline rotkehlchen/assets/resolver.py:142]: You don't need to add it to the cache because if it exists in any of the user global db or packaged globald db then the `resolve_asset` method will store it. It also will raise unknown asset at that point if it doesn't exists

File: rotkehlchen/assets/resolver.py [hunk 12]
```diff
@@ -105,15 +121,31 @@ def resolve_asset_to_class(
...
         )
-        if isinstance(resolved_asset, expected_type) is False:
-            raise WrongAssetType(
+        if isinstance(resolved_asset, expected_type) is True:
+            # resolve_asset returns Asset, but we already narrow type with the if check above
+            return resolved_asset  # type: ignore
+
+        if identifier in CONSTANT_ASSETS:
+            log.debug(f'Attempt to get asset_type for asset with {identifier} using the packaged database')  # noqa: E501
+            resolved_asset = GlobalDBHandler().resolve_asset(
```
Comment (yabirgb) [inline rotkehlchen/assets/resolver.py:137]: This logic is also not needed. The assets is returned by the `resolve_asset` method either from the user globald or the packaged global db

File: rotkehlchen/tests/unit/test_globaldb.py [hunk 13]
```diff
@@ -988,3 +989,27 @@ def test_general_cache(globaldb):
...
+    assert lusd.identifier == A_LUSD.identifier
+
+    # change the type of an asset in `CONSTANT_ASSETS`
+    with globaldb.conn.write_ctx() as cursor:
+        cursor.execute('UPDATE assets SET type=? WHERE identifier=?;', (AssetType.FIAT.serialize_for_db(), A_DAI.identifier))  # noqa: E501
+        assert cursor.rowcount == 1
+    # now resolve and check that no error was raised
+    dai = A_DAI.resolve_to_crypto_asset()
+    assert dai.asset_type == AssetType.EVM_TOKEN
+    assert dai.identifier == A_DAI.identifier
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/test_globaldb.py:1012]: This shouldn't happen. Is not the same problem as if the asset doesn't exists. Here is the user removing it and modifying the globaldb in an unexpected way

**Review by yabirgb — COMMENTED** [review]
Some things are still wrong

File: rotkehlchen/globaldb/handler.py [hunk 14]
```diff
@@ -1783,15 +1807,22 @@ def resolve_asset(identifier: str, form_with_incomplete_data: bool) -> AssetWith
             )
 
     @staticmethod
-    def get_asset_type(identifier: str) -> AssetType:
+    def get_asset_type(
+            identifier: str,
+            use_packaged_db: bool = False,
+    ) -> AssetType:
```
Comment (yabirgb) [inline rotkehlchen/globaldb/handler.py:1813]: This one fits in one line

File: rotkehlchen/tests/unit/test_globaldb.py [hunk 15]
```diff
@@ -988,3 +989,26 @@ def test_general_cache(globaldb):
...
+    assert lusd.asset_type == AssetType.EVM_TOKEN
+    assert lusd.identifier == A_LUSD.identifier
+
+    # delete another asset and try checking its existence
+    with globaldb.conn.write_ctx() as cursor:
+        cursor.execute('DELETE FROM assets WHERE identifier=?;', (A_DAI.identifier, ))  # noqa: E501
+        assert cursor.rowcount == 1
+
+    # now check the asset type is correct and does not raise an error
+    assert A_DAI.is_evm_token()
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/test_globaldb.py:1011]: put the full condition

```suggestion
    # now check the asset type is correct and does not raise an error
    assert A_DAI.is_evm_token() is True
```

File: rotkehlchen/tests/unit/test_globaldb.py [hunk 16]
```diff
@@ -988,3 +989,26 @@ def test_general_cache(globaldb):
...
+
+    with pytest.raises(UnknownAsset):
+        Asset('i-dont-exist').resolve()
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/test_globaldb.py:1014]:
```suggestion
    # check that UnknownAsset is properly raised
    with pytest.raises(UnknownAsset):
        Asset('i-dont-exist').resolve()
```

File: rotkehlchen/globaldb/handler.py (hunk 14 above)
Comment (yabirgb) [inline rotkehlchen/globaldb/handler.py:1813]: I haven't checked this one but probably also fits in one line

File: rotkehlchen/globaldb/handler.py [hunk 17]
```diff
@@ -135,6 +136,22 @@ def __new__(
         GlobalDBHandler.__instance.conn = _initialize_global_db_directory(data_dir, sql_vm_instructions_cb)  # noqa: E501
         return GlobalDBHandler.__instance
 
+    @staticmethod
+    def packaged_db_conn() -> DBConnection:
+        """Return a DBConnection instance for the packaged global db."""
+        if GlobalDBHandler() is not None:
```
Comment (yabirgb) [inline rotkehlchen/globaldb/handler.py:142]: Why did you add this check here? Is not needed. Also for the record if you initialize the globaldb for the first time without the needed arguments it will raise an error

File: rotkehlchen/globaldb/handler.py [hunk 18]
```diff
@@ -135,6 +136,22 @@ def __new__(
...
+                # mypy does not recognize the initialization as that of a singleton
+                return GlobalDBHandler()._packaged_db_conn  # type: ignore
+
+        packaged_db_path = Path(__file__).resolve().parent.parent / 'data' / 'global.db'
+        GlobalDBHandler()._packaged_db_conn = initialize_globaldb(
+            dbpath=packaged_db_path,
+            sql_vm_instructions_cb=DEFAULT_SQL_VM_INSTRUCTIONS_CB,
+        )
+        # mypy does not recognize the initialization as that of a singleton
+        return GlobalDBHandler()._packaged_db_conn  # type: ignore
```
Comment (yabirgb) [inline rotkehlchen/globaldb/handler.py:153]: maybe just to make it clean and avoid the type ignore

```suggestion
        packaged_db_conn = initialize_globaldb(
            dbpath=packaged_db_path,
            sql_vm_instructions_cb=DEFAULT_SQL_VM_INSTRUCTIONS_CB,
        )
        GlobaldbHandler()._packaged_db_conn = packaged_db_conn
        return packaged_db_conn
```

File: rotkehlchen/globaldb/handler.py (hunk 17 above)
Comment (yabirgb) [inline rotkehlchen/globaldb/handler.py:142]: also I don't know the overhead of calling `GlobalDBHandler` each time but maybe you can do

```
globaldb = GlobalDBHandler()
```

once and then use it all the times you call it here

File: rotkehlchen/tests/unit/test_globaldb.py [hunk 19]
```diff
@@ -988,3 +989,26 @@ def test_general_cache(globaldb):
             key_parts=[GeneralCacheType.CURVE_POOL_TOKENS, '123'],
         )
         assert values_8 == values_1
+
+
+def test_packaged_db_check_for_constant_assets(globaldb):
+    """Check that if UnknownAsset is raised for asset in CONSTANT_ASSETS, check the packaged db"""
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/test_globaldb.py:995]:
```suggestion
    """Check that if UnknownAsset is raised for an asset in CONSTANT_ASSETS then the packaged db is queried"""
```

File: rotkehlchen/globaldb/handler.py (hunk 17 above)
Comment (prettyirrelevant) [reply rotkehlchen/globaldb/handler.py:142]: I didn't do that because across the codebase when singleton classes are called, they're not assigned to variables.

Comment (yabirgb) [reply rotkehlchen/globaldb/handler.py:142]: maybe just leave, is true that for example the Inquirer() we call it all the time. I'm not sure either of how bad/good that is for performance

Comment (yabirgb) [reply rotkehlchen/globaldb/handler.py]: It has more overhead that just loading the variable but is true that we do it in other places :/

```
129           0 LOAD_GLOBAL              0 (GlobalDBHandler)
              2 LOAD_ATTR                1 (_GlobalDBHandler__instance)
              4 LOAD_CONST               1 (None)
              6 IS_OP                    1
              8 POP_JUMP_IF_FALSE       16

130          10 LOAD_GLOBAL              0 (GlobalDBHandler)
             12 LOAD_ATTR                1 (_GlobalDBHandler__instance)
             14 RETURN_VALUE
```

File: rotkehlchen/assets/resolver.py [hunk 20]
```diff
@@ -70,10 +75,21 @@ def resolve_asset(
...
-        asset = GlobalDBHandler().resolve_asset(
-            identifier=identifier,
-            form_with_incomplete_data=form_with_incomplete_data,
-        )
+        try:
+            asset = GlobalDBHandler().resolve_asset(
+                identifier=identifier,
+                form_with_incomplete_data=form_with_incomplete_data,
+            )
+        except UnknownAsset:
```
Comment (LefterisJP) [inline rotkehlchen/assets/resolver.py:83]: Didn't I already write it guys? I wrote it before, please for the love of god read my comments and also understand our code.

https://github.com/rotki/rotki/pull/5063#discussion_r1009401454

> It should only happen when the asset is in the list of constant assets and an unknown asset or a wrong asset type exception is raised

You are ignoring an entire category of possible errors here

```suggestion
        except (UnknownAsset, WrongAssetType):
```

File: rotkehlchen/assets/resolver.py [hunk 21]
```diff
@@ -70,10 +75,21 @@ def resolve_asset(
...
+            if identifier not in CONSTANT_ASSETS:
+                raise
+
+            log.debug(f'Attempt to resolve asset with {identifier} using the packaged database')
```
Comment (LefterisJP) [inline rotkehlchen/assets/resolver.py:87]:
```suggestion
            log.debug(f'Attempt to resolve asset {identifier} using the packaged database')
```

File: rotkehlchen/assets/resolver.py [hunk 22]
```diff
@@ -82,14 +98,26 @@ def resolve_asset(
...
 
         instance = AssetResolver()
         cached_data = instance.types_cache.get(identifier)
         if cached_data is not None:
             return cached_data
 
-        asset_type = GlobalDBHandler().get_asset_type(identifier)
+        try:
+            asset_type = GlobalDBHandler().get_asset_type(identifier)
+        except UnknownAsset:
```
Comment (LefterisJP) [inline rotkehlchen/assets/resolver.py:111]: same here

File: rotkehlchen/assets/resolver.py [hunk 23]
```diff
@@ -82,14 +98,26 @@ def resolve_asset(
...
+            if identifier not in CONSTANT_ASSETS:
+                raise
+
+            log.debug(f'Attempt to get asset_type for asset with {identifier} using the packaged database')  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/assets/resolver.py:115]:
```suggestion
            log.debug(f'Attempt to get asset_type for {identifier} using the packaged database')  # noqa: E501
```

A general tip: When you write for `asset with {identifier}` this will render: `for asset with 42432324324` ... which makes no sense as you don't see what that value is. Since by default we use identifiers to describe them in logs you don't need to write it.

File: rotkehlchen/assets/resolver.py [hunk 24]
```diff
@@ -62,6 +66,7 @@ def resolve_asset(
         """
         # TODO: This is ugly here but is here to avoid a cyclic import in the Assets file
         # Couldn't find a reorg that solves this cyclic import
+        from rotkehlchen.constants.assets import CONSTANT_ASSETS  # pylint: disable=import-outside-toplevel  # isort:skip  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/assets/resolver.py:69]: we **really** need to fix this

File: rotkehlchen/constants/assets.py [hunk 25]
```diff
@@ -260,3 +260,5 @@
 A_ALETH = Asset('eip155:1/erc20:0x0100546F2cD4C9D97f798fFC9755E47865FF7Ee6')
 A_STETH = Asset('eip155:1/erc20:0xae7ab96520DE3A18E5e111B5EaAb095312D7fE84')
 A_SAND = Asset('eip155:1/erc20:0x3845badAde8e6dFF049820680d1F14bD3903a5d0')
+
+CONSTANT_ASSETS = {var_data for var_name, var_data in locals().items() if var_name.startswith('A_')}  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/constants/assets.py:264]: why do you need the if? It's all the locals here. Could also have reintrodued the create constants assets script but I guess it's not needed and would be an overkill

File: rotkehlchen/globaldb/handler.py [hunk 26]
```diff
@@ -1750,7 +1772,8 @@ def resolve_asset(identifier: str, form_with_incomplete_data: bool) -> AssetWith
         UNION ALL
         SELECT A.identifier, A.type, null, null, A.name, null, null, null, null, null, null, null, null, null, B.notes, B.type FROM assets AS A JOIN custom_assets AS B on A.identifier=B.identifier WHERE A.identifier = ?
         """  # noqa: E501
-        with GlobalDBHandler().conn.read_ctx() as cursor:
+        connection = GlobalDBHandler().packaged_db_conn() if use_packaged_db is True else GlobalDBHandler().conn  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/globaldb/handler.py:1775]: no this is not the only thing you need to do. 

If you succesfully resolve the asset from the packaged DB then since it's missing from the normal global DB you have to write it there too. You need code to do this.

And code should work for both `UnknownAsset` and `WrongAssetType`. Which mean you should delete the asset if it existed and then replace it with the global DB version.

If you don't do that do you realize what happens? Every single time you resolve the assets you will be going towards the global DB and would have chains of resolutions. Which is crazy

File: rotkehlchen/globaldb/handler.py [hunk 27]
```diff
@@ -1783,15 +1806,19 @@ def resolve_asset(identifier: str, form_with_incomplete_data: bool) -> AssetWith
...
+        If `use_packaged_db` is True, it checks the packaged global db.
+
         May raise:
         - UnknownAsset: if the asset is not present in the database
         """
         if identifier.startswith(NFT_DIRECTIVE):
             return AssetType.NFT
-        with GlobalDBHandler().conn.read_ctx() as cursor:
+
+        connection = GlobalDBHandler().packaged_db_conn() if use_packaged_db is True else GlobalDBHandler().conn  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/globaldb/handler.py:1820]: same here

File: rotkehlchen/tests/unit/test_globaldb.py [hunk 28]
```diff
@@ -988,3 +989,31 @@ def test_general_cache(globaldb):
             key_parts=[GeneralCacheType.CURVE_POOL_TOKENS, '123'],
         )
         assert values_8 == values_1
+
+
+def test_packaged_db_check_for_constant_assets(globaldb):
+    """
+    Check if UnknownAsset is raised for an asset in CONSTANT_ASSETS, the packaged global db
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_globaldb.py:996]: in any tests you write, you have to also check for `WrongAssetType` error and you also have to make sure that later it has also been written in the actual global DB, what I asked you in the previous file

**Review by LefterisJP — COMMENTED** [review]
A lot of comments. Many things off and missing

File: rotkehlchen/constants/assets.py (hunk 25 above)
Comment (yabirgb) [reply rotkehlchen/constants/assets.py:264]: the `if` is needed because locals have other things in here like `__name__`, `__doc__`, `__package__`...

File: rotkehlchen/assets/resolver.py (hunk 20 above)
Comment (yabirgb) [reply rotkehlchen/assets/resolver.py:83]: `resolve_asset` can only raise `WrongAssetType` if the user went to the database and modified the column value for type unless I'm wrong https://github.com/rotki/rotki/blob/0fa9a7cc2dee688db0621750a31b2c6c49c31108/rotkehlchen/globaldb/handler.py#L1770

File: rotkehlchen/assets/resolver.py (hunk 22 above)
Comment (yabirgb) [reply rotkehlchen/assets/resolver.py:111]: `get_asset_type` doesn't raise `WrongAssetType` https://github.com/rotki/rotki/blob/0fa9a7cc2dee688db0621750a31b2c6c49c31108/rotkehlchen/globaldb/handler.py#L1786

Comment (LefterisJP) [reply rotkehlchen/assets/resolver.py:111]: okay fair enough. This one can be ignored then.

File: rotkehlchen/globaldb/handler.py (hunk 26 above)
Comment (yabirgb) [reply rotkehlchen/globaldb/handler.py:1775]: If the type is modified this logic is never reached since is only triggered on `UnknownAsset`

Comment (LefterisJP) [reply rotkehlchen/globaldb/handler.py:1775]: I don't understand what you mean. resolve_asset() can raise `WrongAssetType` which would get us to the packaged DB logic

File: rotkehlchen/assets/resolver.py (hunk 20 above)
Comment (prettyirrelevant) [reply rotkehlchen/assets/resolver.py:83]: `GlobalDBHandler().resolve_asset()` cannot raise `WrongAssetType`. The only place that error is raised is in 

https://github.com/rotki/rotki/blob/0fa9a7cc2dee688db0621750a31b2c6c49c31108/rotkehlchen/assets/resolver.py#L113

**Review by LefterisJP — COMMENTED** [review]
did not do what I asked

**Review by LefterisJP — COMMENTED** [review]
The frontend tests seem to be consistently faling for something

File: rotkehlchen/globaldb/handler.py [hunk 29]
```diff
@@ -1783,15 +1808,68 @@ def resolve_asset(identifier: str, form_with_incomplete_data: bool) -> AssetWith
...
+
+        asset_data: Union[EvmToken, Dict[str, Any]]
+        if asset.asset_type == AssetType.EVM_TOKEN:
+            asset_data = cast(EvmToken, asset)
+        else:
+            asset_data = {
+                'identifier': asset.identifier,
+                'asset_type': asset.asset_type,
+                'forked': getattr(asset, 'forked', None),
+            }
```
Comment (yabirgb) [inline rotkehlchen/globaldb/handler.py:1841]: As you have changed it now you are missing information when adding/editing the asset

**Review by LefterisJP — APPROVED** [review]
LGTM
